crates/consensus/beacon/src/engine/mod.rs

```diff
@@ -716,16 +716,28 @@ where
         {
             Some(parent_hash)
         } else {
-            // TODO: attempt to iterate over ancestors in the invalid cache
+            // iterate over ancestors in the invalid cache
             // until we encounter the first valid ancestor
+            let mut current_hash = parent_hash;
+            let mut current_header = self.invalid_headers.get(&current_hash);
+            while let Some(header) = current_header {
+                current_hash = header.parent_hash;
+                current_header = self.invalid_headers.get(&current_hash);
+
+                // If current_header is None, then the current_hash does not have an invalid
+                // ancestor in the cache
+                if current_header.is_none() {
+                    return Some(current_hash);
```

Comment:
parent hash not being in the invalid headers cache does not necessarily make it valid. it might mean that we simply haven't seen it. you have to additionally check its presence in the blockchain tree and return `None` if it's not there